src/graph/KokkosGraph_CoarsenHeuristics.hpp

```diff
@@ -0,0 +1,1189 @@
+                is_max = true;
+              } else if (tuple_state(v) == max_state) {
+                if (tuple_degree(v) > max_degree) {
+                  is_max = true;
+                } else if (tuple_degree(v) == max_degree) {
+                  if (tuple_idx(v) > max_idx) {
+                    is_max = true;
+                  }
+                }
+              }
+              // pretend edges between two vertices exceeding threshold do not
+              // exist
+              if (degrees(u) > threshold && degrees(v) > threshold) {
+                is_max = false;
+              }
+              if (is_max) {
+                max_state  = tuple_state(v);
+                max_degree = tuple_degree(v);
+                max_idx    = tuple_idx(v);
+              }
+            }
+            tuple_state_update(u)  = max_state;
+            tuple_degree_update(u) = max_degree;
+            tuple_idx_update(u)    = max_idx;
+          });
+
+      Kokkos::parallel_for(
+          policy_t(0, unassigned_total), KOKKOS_LAMBDA(const ordinal_t i) {
+            ordinal_t u     = unassigned(i);
+            tuple_state(u)  = tuple_state_update(u);
+            tuple_degree(u) = tuple_degree_update(u);
+            tuple_idx(u)    = tuple_idx_update(u);
+          });
+
+      ordinal_t next_unassigned_total = 0;
+      Kokkos::parallel_reduce(
+          policy_t(0, unassigned_total),
+          KOKKOS_LAMBDA(const ordinal_t i, ordinal_t& thread_sum) {
+            ordinal_t u = unassigned(i);
+            if (state(u) == 0) {
+              if (tuple_idx(u) == u) {
+                state(u) = 1;
+              }
+              // check if at least one of neighbors are in the IS or will be
+              // placed into the IS
+              else if (tuple_state(u) == 1 ||
+                       tuple_idx(tuple_idx(u)) == tuple_idx(u)) {
+                state(u) = -1;
+              }
+            }
+            if (state(u) == 0) {
+              thread_sum++;
+            }
+          },
+          next_unassigned_total);
+
+      vtx_view_t next_unassigned("next unassigned", next_unassigned_total);
+      Kokkos::parallel_scan(
+          "create next unassigned", policy_t(0, unassigned_total),
+          KOKKOS_LAMBDA(const ordinal_t i, ordinal_t& update,
+                        const bool final) {
+            ordinal_t u = unassigned(i);
+            if (state(u) == 0) {
+              if (final) {
+                next_unassigned(update) = u;
+              }
+              update++;
+            }
+          });
+      unassigned_total = next_unassigned_total;
+      unassigned       = next_unassigned;
+    }
+    return state;
+  }
+
+  static matrix_t coarsen_mis_2(const matrix_t& g) {
+    ordinal_t n = g.numRows();
+
+    typename matrix_t::staticcrsgraph_type::entries_type::non_const_value_type
+        nc           = 0;
+    vtx_view_t vcmap = KokkosGraph::Experimental::graph_mis2_coarsen<
```

Comment:
For this, I recommend using ``graph_mis2_aggregate`` instead. It has the same interface but will give a coarsening with less fill in. It's only a little bit slower.